### PR TITLE
Support extension and scope resource in the same resource group

### PR DESCRIPTION
**Pull request by fengzhou-msft**
# Description
## Introduction
There are 3 kinds of possible paths in the same operation group:
1. Paths with a {scope} variable
  `/{scope}/providers/Microsoft.Foo/bar/{barName}`
2. Paths at a specific scope

- tenant: 
  `/providers/Microsoft.Foo/bar/{barName}`
- management group:
  `/providers/Microsoft.Management/managementGroups/{managementGroupId}/providers/Microsoft.Foo/bar/{barName}`
- subscription: 
  `/subscriptions/{subscriptionId}/providers/Microsoft.Foo/bar/{barName}`
- resource group:
  `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Foo/bar/{barName}`
- resource:
`/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{parentResourcePath}/{resourceType}/{resourceName}/providers/Microsoft.Foo/bar/{barName}`

3. Paths with a single resource id path-parameter
`/{resourceId}`

A `scope resource` must have path 1 and can optionally have path 2 and 3.
An `extension resource` in the same operation group must have muliple different scope paths in 2 and do not have path 1 or 3. Do not confuse it with extension resources in different operation groups  and this PR does not take care of the latter case as it's handled quite differently.

This PR supports generating a single resource container and operation which can take different types of Id and call the same interface to operate on the resources at different scopes. For instance, you can do:
```
managementGroup.GetPolicyAssignmentContainer().Get("managementGroupPA");
subscription.GetPolicyAssignmentContainer().Get("subscriptionPA");
resourceGroup.GetPolicyAssignmentContainer().Get("resourceGroupPA");
```

Methods for path 3 (CreateById/GetById/DeleteById) will be generated but not tested and for Preview, they are removed with readme config as there are alternative methods to achieve similar functionality.

## Code changes
This PR made the following changes:

1. Provide methods to get data (its/parent/ancestor resource type) of a single operation instead of the whole operation group.
2. Make subscriptionId a parameter in method instead of in constructor when the operation group have a tenant ancestor resource type, which meand the operation group not only has subscription scope path, but also management group/tenant scope paths.
3. Merge create/get/delete/list methods for extension/scope resource when they have multiple methods at different scopes. If scope resource has XXAtScope method, then it will be used instead.
3. Support merge methods of different scopes into one method with readme config. Example:
    ```
    merge-operations:
      WhatIf: Deployments_WhatIf_POST;Deployments_WhatIfAtTenantScope_POST;Deployments_WhatIfAtManagementGroupScope_POST;Deployments_WhatIfAtSubscriptionScope_POST
    ```
4. Add ManagementGroupExtensionWriter to add GetResourceContainer methods.
5. ScopeDetection for both OperationGroup and Operation to detect if it contains a path starting with `/{scope}`.
6. Provide IsParentResourceTypeTenant()/IsAncestorResourceTypeTenant() for operation group to include readme config of parent while IsTenantResource()/IsAncestorTenant() only return true for real tenant resource. The new methods can be viewed as equivalent of checking whether the Id of the container is TenantResourceIdentifier.
7. Fix BuildParameterMapping and ShouldPassThrough in some cases and consistently use BuildAndWriteParameter() in all cases instead of somewhere using GetParametersName().
8. Update ResourceContainer/ResourceOperation model to add lists to hold put/get methods at different scopes when it's an extension/scope resource. Otherwise, these lists fallback to hold a single item of createMethod/getMethod.
9. Add MgmtScopeResource and MgmtExtensionResource tests.
10. Refactor WriteXXXMethod to extract common code into a new function for reusability.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs [hunk 1]
```diff
@@ -64,8 +67,17 @@ private string GetRequestParameterName (RequestParameter requestParameter)
...
+            var isAncestorResourceTypeTenant = false;
+            try
+            {
+                isAncestorResourceTypeTenant = _operationGroup.IsAncestorResourceTypeTenant(_context);
+            }
+            catch (ArgumentException)
+            {
+                // do nothing for exception in dataplane
+            }
+            List<RequestParameter> parameters = (isAncestorResourceTypeTenant ? operation.Parameters.OrderBy(p => !p.Language.Default.Name.Equals("subscriptionId", StringComparison.InvariantCultureIgnoreCase) && !p.Language.Default.Name.Equals("managementGroupId", StringComparison.InvariantCultureIgnoreCase)).ToList() : operation.Parameters).Concat(requestParameters).ToList();
```
Comment (fengzhou-msft) [inline src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:80]: The workaround can be removed after #1423 is merged.

Comment (m-nash) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:80]: This pr should have covered it https://github.com/Azure/autorest.csharp/pull/1431 instead of 1423

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs [hunk 2]
```diff
@@ -62,5 +62,10 @@ public List<Parameter> PathParameters
                 ).ToList();
             }
         }
+
+        public bool IsByIdMethod()
+        {
+            return this.Operation?.Requests.FirstOrDefault()?.Protocol.Http is HttpRequest httpRequest && this.Parameters.Count() > 0 && $"/{{{this.Parameters[0].Name}}}".Equals(httpRequest.Path);
+        }
```
Comment (ShivangiReja) [inline src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs:69]: Can we remove this management sdk specific method from common to somewhere inside `AutoRest.CSharp/Mgmt/`?

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtConfiguration.cs [hunk 3]
```diff
@@ -17,11 +17,22 @@ public class MgmtConfiguration
             string[] singletonResource,
             JsonElement? operationGroupToResourceType = default,
             JsonElement? operationGroupToResource = default,
-            JsonElement? operationGroupToParent = default)
+            JsonElement? operationGroupToParent = default,
+            JsonElement? mergeOperations = default)
```
Comment (ShivangiReja) [inline src/AutoRest.CSharp/Mgmt/AutoRest/MgmtConfiguration.cs:21]: Can we add some examples/tests for mergeOperations in the readme?

Comment (ShivangiReja) [inline src/AutoRest.CSharp/Mgmt/AutoRest/MgmtConfiguration.cs:21]: I don't see this config in the readme.md. Can we add some examples/tests for mergeOperations in the readme?

Comment (fengzhou-msft) [reply src/AutoRest.CSharp/Mgmt/AutoRest/MgmtConfiguration.cs:21]: It would take a while to extract the Deployments resource which uses mergeOperations from resources.json as I need to remove definitions already in Core. I will add the tests in a separate PR.

File: src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs (hunk 2 above)
Comment (fengzhou-msft) [reply src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs]: Good catch. Moved to `MethodExtensions.cs`.

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs [hunk 4]
```diff
@@ -64,8 +67,17 @@ private string GetRequestParameterName (RequestParameter requestParameter)
...
+            //TODO: remove this part of code which makes subscriptionId and managementGroupId as the first parameter when parameters are sorted by their order in path.
+            var isAncestorResourceTypeTenant = false;
+            try
+            {
+                isAncestorResourceTypeTenant = _operationGroup.IsAncestorResourceTypeTenant(_context);
+            }
+            catch (ArgumentException)
+            {
+                // do nothing for exception in dataplane
+            }
```
[the remark this answers is not in the document]
Comment (fengzhou-msft) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs]: This part of code is no longer needed and removed.

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs (hunk 1 above)
Comment (fengzhou-msft) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs]: Code removed.

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtTarget.cs [hunk 5]
```diff
@@ -125,6 +125,14 @@ public static void Execute(GeneratedCodeWorkspace project, CodeModel codeModel,
             var subscriptionExtensionsCodeWriter = new CodeWriter();
             subscriptionExtensionsWriter.WriteExtension(subscriptionExtensionsCodeWriter, context);
             project.AddGeneratedFile($"Extensions/{ResourceTypeBuilder.TypeToExtensionName[ResourceTypeBuilder.Subscriptions]}.cs", subscriptionExtensionsCodeWriter.ToString());
+
+            if (context.Library.ResourceContainers.Any(c => c.OperationGroup.Operations.Any(op => op.ParentResourceType() == ResourceTypeBuilder.ManagementGroups)))
```
[the remark this answers is not in the document]
Comment (fengzhou-msft) [reply src/AutoRest.CSharp/Mgmt/AutoRest/MgmtTarget.cs]: Added.

File: test/AutoRest.TestServer.Tests/Mgmt/OutputLibrary/MgmtScopeResourceTest.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+
+using AutoRest.CSharp.Mgmt.Decorator;
+using NUnit.Framework;
+
+namespace AutoRest.TestServer.Tests.Mgmt.OutputLibrary
+{
+    internal class MgmtScopeResourceTests : OutputLibraryTestBase
```
[the remark this answers is not in the document]
Comment (fengzhou-msft) [reply test/AutoRest.TestServer.Tests/Mgmt/OutputLibrary/MgmtScopeResourceTest.cs:9]: Added.

File: test/AutoRest.TestServer.Tests/Mgmt/OutputLibrary/MgmtExtensionResourceTests.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+
+using AutoRest.CSharp.Mgmt.Decorator;
+using NUnit.Framework;
+
+namespace AutoRest.TestServer.Tests.Mgmt.OutputLibrary
+{
+    internal class MgmtExtensionResourceTests : OutputLibraryTestBase
```
[the remark this answers is not in the document]
Comment (fengzhou-msft) [reply test/AutoRest.TestServer.Tests/Mgmt/OutputLibrary/MgmtExtensionResourceTests.cs:9]: Added.

File: src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs [hunk 8]
```diff
@@ -62,5 +62,6 @@ public List<Parameter> PathParameters
                 ).ToList();
             }
         }
+
```
Comment (ShivangiReja) [inline src/AutoRest.CSharp/Common/Output/Models/Requests/RestClientMethod.cs:65]: nit: extra line :)

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs [hunk 9]
```diff
@@ -64,7 +64,6 @@ private string GetRequestParameterName (RequestParameter requestParameter)
         public RestClientMethod BuildMethod(Operation operation, HttpRequest httpRequest, IEnumerable<RequestParameter> requestParameters, DataPlaneResponseHeaderGroupType? responseHeaderModel, string accessibility)
         {
             Dictionary<RequestParameter, ConstructedParameter> allParameters = new();
-
```
Comment (ShivangiReja) [inline src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:67]: nit: we can undo the changes in this file.

File: src/AutoRest.CSharp/Mgmt/Generation/MgmtClientBaseWriter.cs [hunk 10]
```diff
@@ -354,15 +520,8 @@ protected void WriteClientMethod(CodeWriter writer, ClientMethod clientMethod, s
                     {
                         writer.Append($"await ");
                     }
-
-                    var parameterNames = GetParametersName(restClientMethod, operationGroup, context);
                     writer.Append($"{restClientName ?? RestClientField}.{CreateMethodName(restClientMethod.Name, async)}(");
-                    // TODO -- we need to change this to BuildAndWriteParameters(writer, clientMethod) to make it be able to handle more cases
-                    // but directly replace the following logic by this function is causing issues
```
Comment (fengzhou-msft) [inline src/AutoRest.CSharp/Mgmt/Generation/MgmtClientBaseWriter.cs:361]: @ArcturusZhang I've made some fix in BuildParameterMapping and ShouldPassThrough, then using BuildAndWriteParameters here looks fine from the test eaxmples and actually fixed some parameter value expressions in MgmtListMethods. Please let me know if you have any concerns.

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs [hunk 11]
```diff
@@ -64,8 +64,8 @@ private string GetRequestParameterName (RequestParameter requestParameter)
...
             List<RequestParameter> parameters = operation.Parameters.Concat(requestParameters).ToList();
+
             // Remove ignored headers
```
Comment (fengzhou-msft) [inline src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:69]:
```suggestion

            List<RequestParameter> parameters = operation.Parameters.Concat(requestParameters).ToList();
            // Remove ignored headers
```

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtConfiguration.cs (hunk 3 above)
Comment (fengzhou-msft) [reply src/AutoRest.CSharp/Mgmt/AutoRest/MgmtConfiguration.cs:21]: Example added in readme of MgmtScopeResource.

File: src/AutoRest.CSharp/Mgmt/Output/ResourceOperation.cs [hunk 12]
```diff
@@ -91,7 +96,12 @@ internal class ResourceOperation : TypeProvider
 
         public IEnumerable<ResourceListMethod>? ManagementGroupExtensionListMethods => _managementGroupExtensionsListMethods ??= GetManagementGroupExtensionsListMethods();
 
-        public virtual ClientMethod? GetMethod => _getMethod ??= Methods.FirstOrDefault(m => m.Name.StartsWith("Get") && m.RestClientMethod.Responses[0].ResponseBody?.Type.Name == ResourceData.Type.Name);
+        public virtual ClientMethod? GetMethod => _getMethod ??= Methods.FirstOrDefault(m => IsGetResourceMethod(m) && m.RestClientMethod.Parameters.FirstOrDefault()?.Name.Equals("scope") == true) ?? Methods.OrderBy(m => m.Name.Length).FirstOrDefault(m => IsGetResourceMethod(m));
```
Comment (m-nash) [inline src/AutoRest.CSharp/Mgmt/Output/ResourceOperation.cs:99]: Can we find a way to share the code from here with https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Mgmt/Output/MgmtRestClient.cs#L27?

**Review by m-nash — CHANGES_REQUESTED** [review]

**Review by m-nash — APPROVED** [review]
